src/crypto/hpke/mod.rs

```diff
@@ -0,0 +1,236 @@
+//! Provides decryption primitives for HPKE according to the [`specification`].
+//!
+//! [`specification`]: https://github.com/patcg-individual-drafts/ipa/pull/31
+
+use hpke::aead::AeadTag;
+use hpke::{single_shot_open_in_place_detached, Deserializable, OpModeR};
+use std::{fmt, io};
+
+mod aad;
+mod registry;
+
+pub use aad::AssociatedData;
+pub use registry::KeyRegistry;
+
+/// IPA ciphersuite
+type IpaKem = hpke::kem::X25519HkdfSha256;
+type IpaAead = hpke::aead::AesGcm128;
+type IpaKdf = hpke::kdf::HkdfSha256;
+
+pub type KeyIdentifier = u8;
+pub type Epoch = u16;
+type IpaPublicKey = <IpaKem as hpke::kem::Kem>::PublicKey;
+type IpaPrivateKey = <IpaKem as hpke::kem::Kem>::PrivateKey;
+
+#[derive(Debug)]
+pub struct BottomError;
+
+impl From<hpke::HpkeError> for BottomError {
+    fn from(_value: hpke::HpkeError) -> Self {
+        Self
+    }
+}
+
+impl From<io::Error> for BottomError {
+    fn from(_value: io::Error) -> Self {
+        Self
+    }
+}
+
+impl fmt::Display for BottomError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Failed to open ciphertext")
+    }
+}
+
+/// Opens ciphertext in place by first obtaining the secret key from [`key_registry`]
+/// using epoch and key from [`aad`] and then applying [`HPKE decryption`] to the provided ciphertext.
+///
+/// ## Errors
+/// If ciphertext cannot be opened for any reason.
+///
+/// [`HPKE decryption`]: https://datatracker.ietf.org/doc/html/rfc9180#name-encryption-and-decryption
+pub fn open_in_place(
+    key_registry: &KeyRegistry,
+    encap_key: &[u8],
+    ciphertext: &mut [u8],
+    tag: &[u8],
+    aad: &AssociatedData,
+) -> Result<(), BottomError> {
+    //TODO: log errors, but don't return them
+    let info = aad.to_bytes();
+    let encap_key = <IpaKem as hpke::Kem>::EncappedKey::from_bytes(encap_key)?;
+    let tag = AeadTag::<IpaAead>::from_bytes(tag)?;
+    let sk = key_registry.private_key(aad.epoch(), aad.key_id());
+
+    Ok(single_shot_open_in_place_detached::<_, IpaKdf, IpaKem>(
+        &OpModeR::Base,
+        sk,
+        &encap_key,
+        &info,
+        ciphertext,
+        &[],
+        &tag,
+    )?)
+}
+
+#[cfg(all(test, not(feature = "shuttle")))]
+mod tests {
+    use super::*;
+    use crate::secret_sharing::XorReplicated;
+
+    use hpke::{single_shot_seal_in_place_detached, OpModeS, Serializable};
+    use rand::rngs::StdRng;
+    use rand_core::{CryptoRng, RngCore, SeedableRng};
+
+    use crate::crypto::hpke::aad::{HelperOrigin, SiteOrigin};
+
+    #[derive(Clone)]
+    struct MatchKeyEncryption {
```

Comment:
It strikes me that we might want a way to represent the thing that we parse off the wire.  Maybe this isn't it (we might want to reference a common `info`), but it might be worth thinking that through a little, to make decrypting lots of stuff easier.